### PR TITLE
Cross-compile builder to Windows for PRs on Travis

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
I chose a completely arbitrary windows target here (I have no idea what's best, we could do multiple -- they are relatively fast).

r? @alexcrichton

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Oh, and I should cc @aidanhs and probably @kennytm

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/364922471) ([raw log](https://api.travis-ci.org/v3/job/364922471/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:07] error: Server does not allow request for unadvertised object e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f
[00:00:07] Fetched in submodule path 'src/libcompiler_builtins', but it did not contain e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f. Direct fetching of that commit failed.
---
[00:00:08] error: Server does not allow request for unadvertised object e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f
[00:00:08] Fetched in submodule path 'src/libcompiler_builtins', but it did not contain e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f. Direct fetching of that commit failed.
---
[00:00:09] error: Server does not allow request for unadvertised object e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f
[00:00:09] Fetched in submodule path 'src/libcompiler_builtins', but it did not contain e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f. Direct fetching of that commit failed.
---
[00:00:10] error: Server does not allow request for unadvertised object e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f
[00:00:10] Fetched in submodule path 'src/libcompiler_builtins', but it did not contain e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f. Direct fetching of that commit failed.
---
[00:00:11] error: Server does not allow request for unadvertised object e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f
[00:00:11] Fetched in submodule path 'src/libcompiler_builtins', but it did not contain e0bb71ae93695d01cbeaf5dd7a43f0a7b40fb32f. Direct fetching of that commit failed.
[00:00:11] The command has failed after 5 attempts.
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:0fce8196:start=1523416322650234846,finish=1523416322656769057,duration=6534211
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:2b17b0e2
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:2b17b0e2:start=1523416322663415776,finish=1523416322669924092,duration=6508316
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:059b9345
$ dmesg | grep -i kill
[   10.607272] init: failsafe main process (1093) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

**Review by alexcrichton — COMMENTED** [review]
Nice! 

Looking at Travis it looks like this took ~6m, although I don't think it's checking librustc yet. I wonder if to solve the miniz-sys issue we could create a shim shell script C compiler which basically does nothing? Or otherwise provide overrides for the relevant build scripts?

Additionally, would it be possible to skip the submodule updates in this PR? I think most of them shouldn't be necessary other than maybe libcompiler_builtins, right?

**Comment by Mark-Simulacrum** [comment]
We need submodule updates for at least `src/tools/cargo` and `src/tools/rls` I think (and maybe others) since Cargo needs those to resolve the Cargo.lock/Cargo.toml configuration IIRC.

**Comment by alexcrichton** [comment]
Ah good point yeah, it may be that the submodules are too entangled to remove them from being updated

**Comment by Mark-Simulacrum** [comment]
Note: Making all build scripts no-ops doesn't work as winapi (at least) needs it's build script to run in order to compile (or so I assume). That means we'll probably need a whitelist -- how do you feel about that, @alexcrichton? Is a whitelist fine? I'm not really happy with that solution/answer...

**Comment by alexcrichton** [comment]
Hm ok we probably don't want to bend over backwards too much as this may run the risk of causing more bugs than it discovers, so maybe we could use a docker image with a MinGW toolchain and go from there?

The main thing we want to skip is LLVM mostly

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/365189161) ([raw log](https://api.travis-ci.org/v3/job/365189161/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:00c61524:start=1523462589305727837,finish=1523462589311236929,duration=5509092
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:013423bb
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:013423bb:start=1523462589316487295,finish=1523462589321653722,duration=5166427
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:062f7100
$ dmesg | grep -i kill
[   11.259908] init: failsafe main process (1094) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/365196817) ([raw log](https://api.travis-ci.org/v3/job/365196817/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:111f63b0:start=1523463323129364932,finish=1523463323139017390,duration=9652458
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:33269af8
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:33269af8:start=1523463323147430575,finish=1523463323156014950,duration=8584375
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:3370a7f0
$ dmesg | grep -i kill
[   11.520591] init: failsafe main process (1094) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

**Comment by Mark-Simulacrum** [comment]
@alexcrichton Okay, so I've gone with the docker approach now and it compiles std, test, and rustc. I think this is pretty much ready to land from my perspective pending any concerns from you; there are effectively no rustbuild changes to speak of: just making rustc get built on check for --target (not --host) builds.

File: .travis.yml [hunk 1]
```diff
@@ -176,6 +176,8 @@ matrix:
       if: branch = auto
     - env: IMAGE=x86_64-gnu-distcheck
       if: branch = auto
+    - env: IMAGE=mingw-check
+      if: type = pull_request
```
Comment (alexcrichton) [inline .travis.yml]: I'd actually run this on the `auto` branch as well just to be extra sure it's always passing

File: src/ci/docker/mingw-check/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  gdb \
+  xz-utils \
+  libssl-dev \
+  pkg-config \
+  mingw-w64
+
+COPY scripts/sccache.sh /scripts/
+RUN sh /scripts/sccache.sh
+
+ENV SCRIPT python2.7 ../x.py check --target=i686-pc-windows-gnu -vv
```
Comment (alexcrichton) [inline src/ci/docker/mingw-check/Dockerfile]: The `-vv` here can probably be removed

File: src/bootstrap/check.rs [hunk 3]
```diff
@@ -64,7 +64,6 @@ pub struct Rustc {
 
 impl Step for Rustc {
     type Output = ();
-    const ONLY_HOSTS: bool = true;
```
Comment (alexcrichton) [inline src/bootstrap/check.rs]: Instead of removing this, would `--host=i686-pc-windows-gnu` work below?

**Review by alexcrichton — COMMENTED** [review]
This looks great! Just a few minor nits

**Comment by Mark-Simulacrum** [comment]
Nits fixed.

**Pull request merged**